Restructure CLI Commands

dnephin opened the pull request:
- Fixes #13509
- Fixes #8829
- Fixes #8756


- create a new `docker image` and `docker container` command
- add the image and container commands as subcommands of the new command groups with a few commands renamed:
  - `images` -> `image ls` (with aliases)
  - `rmi` -> `image rm` (with aliases)
  - `ps` -> `container ls` (with aliases)
- add `image inspect` and `container inspect` to match the other command groups
- add 3-letter allises for all commands groups
  - `docker con`
  - `docker img`
  - `docker srv`
  - `docker net`
  - `docker vol`

dnephin committed:
Move canonical image and container commands into a command group

Hide some top level commands
Add docker container and image inspect commands.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

icecrime commented:
Absolutely LGTM on design, :tada:, :100:, etc.

My only nit, but I won't fight for it, is that I'm skeptical about the need for short aliases (`docker con`, `docker vol`, etc.).

nathanleclaire commented:
> `docker con`

![image](https://cloud.githubusercontent.com/assets/1476820/17989771/a32378d4-6ae4-11e6-838e-bcc07d60992d.png)

MHBauer commented:
Not big on the short aliases, but I like the thought for everything else.

vdemeester commented:
Not a huge fan of short aliases but other than that, just like @icecrime, absolutely LGTM on design :sunglasses: :tada:

dnephin commented:
My rational for the short aliases is this:

It's going to be difficult to convince people to use `docker container rm` instead of `docker rm`. It's just much longer to type.  Adding the short aliases make the situation a lot better because it's just 3 extra characters, instead of 9!

If we don't make it easy to use the new canonical commands we're never going to be able to remove them.

cpuguy83 commented:
You think we'll be able to remove something like `docker rm` or `docker run`? I doubt that.

dnephin commented:
`docker run` no, we aren't even hiding it, it's staying as a permanent alias.

The rest, not any time soon. Maybe in a year or two if we're good about converting all the docs to the new form, and communicating the new canonical way of doing things.

stevvooe commented:
I love this, but if we do it, we should take this chance to fix the output of inspect to default to  be pretty and composite. This was a HUGE point of contention during the implementation of swarm services. Making `inspect` output text composed of multiple API calls will address this problem and let us really grow the usability of `inspect`, while addressing long standing issues.

stevvooe commented on api/client/image/cmd.go:
```diff
@@ -0,0 +1,38 @@
+		NewPullCommand(dockerCli),
+		NewPushCommand(dockerCli),
+		NewSaveCommand(dockerCli),
+		NewSearchCommand(dockerCli),
```
I don't think we should carry over search to `docker image`. We aren't searching the local image store, we are searching remote registries. Future extension to search won't happen under `docker image`.

icecrime replied:
That's a good point. Maybe it belongs with `login` and `logout` under another top-level command?

dnephin replied:
Makes sense. I'll move the search command file to the correct package, and move it back to the top level.

stevvooe replied:
There may be a `docker registry` command, but that is a whole 'notha can o wyrms!

stevvooe commented on api/client/image/cmd.go:
```diff
@@ -0,0 +1,38 @@
+	cmd.AddCommand(
+		NewBuildCommand(dockerCli),
+		NewHistoryCommand(dockerCli),
+		NewImportCommand(dockerCli),
```
This is a container operation, if I remember correctly.

stevvooe replied:
Nevermind. It is `docker export` that is for containers.

OJFord commented:
> My rational for the short aliases is this:
> It's going to be difficult to convince people to use `docker container rm` instead of `docker rm`. It's just much longer to type.

Realistically people won't do that though, they'll type `docker c<tab> rm`. As long as the completion is the full "container" (i.e. "con" isn't an entry) I'm not bothered, but for this reason I don't think it's necessary.

dnephin committed:
Move the search command to the registry package.

And move it back to the top-level command.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
>  we should take this chance to fix the output of inspect to default to be pretty and composite

I'm not sure that we can change `inspect`, but we could definitely introduce a new command that defaults to pretty print, and hide `inspect` from help. Do you know if there is an issue with a proposal for this change? I'd like to keep that discussion and change separate from this PR.

nathanleclaire commented:
> Realistically people won't do that though, they'll type docker c<tab> rm. As long as the completion is the full "container" (i.e. "con" isn't an entry) it wouldn't bother me, but for this reason I don't think it's necessary.

My 2c -- I'm not really a fan of completion on subcommands, since it requires that you install and maintain additional completion scripts instead of being something that works out of the box.  In `docker container`'s case as well, keep in mind that it's competing with `docker commit` and others for the same completion, so by the time you've typed enough characters to differentiate `container` for completion, you've already typed `docker con` anyway.

I wouldn't really like to see `docker ps` go away pretty much ever since it is such deeply ingrained habit for me, is extremely short and frequently invoked and I don't really mind having porcelain around the "true" commands (git style, i.e. "every command is just a wrapper around rebase").  But I understand the motivation.

I'd definitely like to see more elaboration around where we draw the lines here, e.g., why would it be OK for `docker service scale foo=3` to alias to `docker service update --replicas 3 foo` (porcelain wrapper like mentioned above) but not OK for `docker ps` to alias to `docker container ls`.

stevvooe commented:
> I'm not sure that we can change inspect, but we could definitely introduce a new command that defaults to pretty print, and hide inspect from help. Do you know if there is an issue with a proposal for that change? I'd like to keep that discussion and change separate from this PR.

Not that I know of. File away or ping me again and I'll cook something up.

I agree that this shouldn't be a part of this PR, if we frame it that way.

dnephin commented:
This is now rebased, and no longer hides the commands. It only adds the new canonical commands.

dnephin commented:
rebased and green again

@icecrime what do you think?

(the commands are no longer hidden)

icecrime commented:
@dnephin Thanks, I think this is less disruptive for the user this way. The only unfortunate (but I believe unavoidable) thing is those two very similar commands:

```
  image       Manage Docker images
  images      List images
```

1) Rather than removing the command hiding entirely, I would make it dependent on some environment variable perhaps? Please ignore the choice of the variable name, but for example:

```
diff --git a/cli/command/commands/commands.go b/cli/command/commands/commands.go
index 6d103d6..f972719 100644
--- a/cli/command/commands/commands.go
+++ b/cli/command/commands/commands.go
@@ -1,6 +1,8 @@
 package commands
 
 import (
+       "os"
+
        "github.com/docker/docker/cli/command"
        "github.com/docker/docker/cli/command/container"
        "github.com/docker/docker/cli/command/image"
@@ -73,6 +75,9 @@ func AddCommands(cmd *cobra.Command, dockerCli *command.DockerCli) {
 }
 
 func hide(cmd *cobra.Command) *cobra.Command {
+       if v := os.Getenv("DOCKER_HIDE_LEGACY_COMMANDS"); v == "" {
+               return cmd
+       }
        cmdCopy := *cmd
        cmdCopy.Hidden = true
        cmdCopy.Aliases = []string{}
```

This would give the following:

```
# docker

Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/root/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/root/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/root/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/root/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Commands:
  attach      Attach to a running container
  build       Build an image from a Dockerfile
  commit      Create a new image from a container's changes
  container   Manage Docker containers
  cp          Copy files/folders between a container and the local filesystem
  create      Create a new container
  diff        Inspect changes on a container's filesystem
  events      Get real time events from the server
  exec        Run a command in a running container
  export      Export a container's filesystem as a tar archive
  history     Show the history of an image
  image       Manage Docker images
  images      List images
  import      Import the contents from a tarball to create a filesystem image
  info        Display system-wide information
  inspect     Return low-level information on a container, image or task
  kill        Kill one or more running containers
  load        Load an image from a tar archive or STDIN
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  logs        Fetch the logs of a container
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  pause       Pause all processes within one or more containers
  port        List port mappings or a specific mapping for the container
  ps          List containers
  pull        Pull an image or a repository from a registry
  push        Push an image or a repository to a registry
  rename      Rename a container
  restart     Restart one or more containers
  rm          Remove one or more containers
  rmi         Remove one or more images
  run         Run a command in a new container
  save        Save one or more images to a tar archive (streamed to STDOUT by default)
  search      Search the Docker Hub for images
  service     Manage Docker services
  start       Start one or more stopped containers
  stats       Display a live stream of container(s) resource usage statistics
  stop        Stop one or more running containers
  swarm       Manage Docker Swarm
  tag         Tag an image into a repository
  top         Display the running processes of a container
  unpause     Unpause all processes within one or more containers
  update      Update configuration of one or more containers
  version     Show the Docker version information
  volume      Manage Docker volumes
  wait        Block until one or more containers stop, then print their exit codes

Run 'docker COMMAND --help' for more information on a command.
# DOCKER_HIDE_LEGACY_COMMANDS=1 docker

Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/root/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/root/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/root/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/root/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Commands:
  build       Build an image from a Dockerfile
  container   Manage Docker containers
  events      Get real time events from the server
  image       Manage Docker images
  info        Display system-wide information
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  run         Run a command in a new container
  search      Search the Docker Hub for images
  service     Manage Docker services
  swarm       Manage Docker Swarm
  version     Show the Docker version information
  volume      Manage Docker volumes

Run 'docker COMMAND --help' for more information on a command.
```

2) In the help output, can we group together the "proper" ones from the legacy? Again, please ignore the choice of working, but for example:

```
# DOCKER_HIDE_LEGACY_COMMANDS=1 docker

Usage:  docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/root/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/root/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/root/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/root/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Management commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes

Operations:
  build       Build an image from a Dockerfile
  events      Get real time events from the server
  info        Display system-wide information
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  run         Run a command in a new container
  search      Search the Docker Hub for images
  version     Show the Docker version information
```

dnephin commented:
Sounds good. I'll make those changes

dnephin committed:
Only hide commands if the env variable is set.

Better formatting for usage template.
Group commands in usage to management/operation commands.
Remove the word Docker from the description of management commands.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
I've made those changes, here's what it looks like by default.

```
Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/home/daniel/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/home/daniel/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/home/daniel/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/home/daniel/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Management Commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes

Commands:
  attach      Attach to a running container
  build       Build an image from a Dockerfile
  commit      Create a new image from a container's changes
  cp          Copy files/folders between a container and the local filesystem
  create      Create a new container
  diff        Inspect changes on a container's filesystem
  events      Get real time events from the server
  exec        Run a command in a running container
  export      Export a container's filesystem as a tar archive
  history     Show the history of an image
  images      List images
  import      Import the contents from a tarball to create a filesystem image
  info        Display system-wide information
  inspect     Return low-level information on a container, image or task
  kill        Kill one or more running containers
  load        Load an image from a tar archive or STDIN
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  logs        Fetch the logs of a container
  pause       Pause all processes within one or more containers
  port        List port mappings or a specific mapping for the container
  ps          List containers
  pull        Pull an image or a repository from a registry
  push        Push an image or a repository to a registry
  rename      Rename a container
  restart     Restart one or more containers
  rm          Remove one or more containers
  rmi         Remove one or more images
  run         Run a command in a new container
  save        Save one or more images to a tar archive (streamed to STDOUT by default)
  search      Search the Docker Hub for images
  start       Start one or more stopped containers
  stats       Display a live stream of container(s) resource usage statistics
  stop        Stop one or more running containers
  tag         Tag an image into a repository
  top         Display the running processes of a container
  unpause     Unpause all processes within one or more containers
  update      Update configuration of one or more containers
  version     Show the Docker version information
  wait        Block until one or more containers stop, then print their exit codes

Run 'docker COMMAND --help' for more information on a command.
```

With hidden commands enabled by environment variable
```

Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/home/daniel/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/home/daniel/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/home/daniel/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/home/daniel/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Management Commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes

Commands:
  build       Build an image from a Dockerfile
  events      Get real time events from the server
  info        Display system-wide information
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  run         Run a command in a new container
  search      Search the Docker Hub for images
  version     Show the Docker version information

Run 'docker COMMAND --help' for more information on a command.
```

duglin commented:
I'm sure someone better at UX, or docs, that me could come up with something really smart (I can't right now), but let's brainstorm on this section:
```
Management Commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes
```
The text on the right seems kind of obvious and therefore verbose.

Maybe even just list them, such as:
```
Resource Management Commands:
   container, image, network, node, service, swarm
```
I dunno....

vdemeester commented:
LGTM 🐸
@dnephin needs a rebase 👼

thaJeztah commented:
I agree with @duglin's comment as well.

Also (following the earlier discussion); I think we should remove the three-letter aliases for now (`docker con`, `docker img` etc.). It looks like we're not sure they're really needed, and in case of doubt, I suggest to *not* do it. It's way easier to add them later, then to deprecate.

If you feel strongly about the three-letter aliases, perhaps open a separate PR, so that this PR won't get blocked because of that. (Sorry for my late reply - blame my vacation :smile:)

thaJeztah commented:
Oh, other than that (but I need to give this a more in-depth look); I love this PR!

dnephin commented:
Hmm, I don't get any conflicts when I rebase with `master`.  I'll make those changes.

dnephin commented:
Ok, I've removed the short aliases (I'll submit as a follow up PR for more discussion), and I've removed the word "Docker" from the usage test of the management commands as suggested.

I don't like the idea of listing them on a single line. I think they aren't noticeable enough that way, and they are going to look weird with text wrapping.

icecrime commented:
LGTM again.

mlaventure commented:
LGTM (assuming it's :green_heart:)

thaJeztah commented:
Just discussed with @dnephin that we're gonna do docs (and completion scripts) in a follow-up PR. I'll open a tracking issue

thaJeztah commented:
And, it's green!

vieux commented:
LGTM !

icecrime commented:
:clap:

gittycat commented:
Is there any chance of differentiating between Legacy|old|standalone Swarm and "Swarm Mode" at least in the doc. For instance:

```
Management Commands:
  ...
  node        Manage Docker Swarm mode nodes
  service     Manage Docker services
  swarm       Manage Docker legacy Swarm
  ...
```

The old vs new swarm is a recurring area of misunderstanding over at Stackoverflow and elsewhere. The "swarm mode" should really have been named something else in my opinion. Sorry to bring this up here but since we're stuck with "Swarm" I would love the doc to at least acknowledge that it's dealing with two different technologies and use some naming to help distinguish them.

dnephin commented:
@gittycat  `docker swarm` is not the legacy swarm, it's the new "swarm mode". Only `swarm` is the old swarm, which is not available through this CLI.

stevvooe commented:
@gittycat Rule of thumb: if there is a `docker` in front of your command, you're probably using Swarm-mode.

I usually refer to them as "Docker Services" for clarification.

albers commented:
ping @sdurrheimer there's a lot of work waiting for you here :cold_sweat:, PTAL

sdurrheimer commented:
@albers I see 😲, thx for the ping.

lpalgarvio commented:
kudos!

i was actually using custom make scripts to do just this with almost equal commands!!!

thanks for standardizing in 1.13! saved me from opening a ticket =)

icetech233 commented:
整那么多乱七八糟的，我日你m
